test/function.localstack.test.ts

```diff
@@ -603,7 +548,129 @@ localstackTestSuite("functionStack", (testResource, _stack, _app) => {
     "started!"
   );
 
-  testFunctionResource(
+  test(
+    "tokens",
+    (parent) => {
+      const token = Token.asString("hello");
+      const obj = { iam: "object" };
+      const token2 = Token.asAny(obj);
+      const obj2 = { iam: Token.asString("token") };
+      const nestedToken = Token.asAny(obj2);
+      const numberToken = Token.asNumber(1);
+      const listToken = Token.asList(["1", "2"]);
+      const nestedListToken = Token.asList([Token.asString("hello")]);
+      return new Function(
+        parent,
+        "function",
+        localstackClientConfig,
+        async () => {
+          return {
+            string: token,
+            object: token2 as unknown as typeof obj,
+            nested: nestedToken as unknown as typeof obj2,
+            number: numberToken,
+            list: listToken,
+            nestedList: nestedListToken,
+          };
+        }
+      );
+    },
+    {
+      string: "hello",
+      object: { iam: "object" },
+      nested: { iam: "token" },
+      number: 1,
+      list: ["1", "2"],
+      nestedList: ["hello"],
+    }
+  );
+
+  test(
+    "serialize entire table",
+    (parent) => {
+      const table = new aws_dynamodb.Table(parent, "table", {
+        partitionKey: {
+          name: "key",
+          type: aws_dynamodb.AttributeType.STRING,
+        },
+      });
+      const get = $AWS.DynamoDB.GetItem;
+      table.addGlobalSecondaryIndex({
+        indexName: "testIndex",
+        partitionKey: {
+          name: "key",
+          type: aws_dynamodb.AttributeType.STRING,
+        },
+      });
+
+      const flTable = new Table(table);
+
+      return new Function(
+        parent,
+        "function",
+        localstackClientConfig,
+        async () => {
+          get({
+            TableName: flTable,
+            Key: {
+              key: { S: "hi" },
+            },
+          });
+        }
+      );
+    },
+    null
+  );
+
+  test(
+    "serialize token with nested string",
+    (parent) => {
+      const table = new aws_dynamodb.Table(parent, "table", {
+        partitionKey: {
+          name: "key",
+          type: aws_dynamodb.AttributeType.STRING,
+        },
+      });
+
+      const obj = { key: table.tableArn };
+      const token = Token.asAny(obj);
+
+      return {
+        func: new Function(
+          parent,
+          "function",
+          localstackClientConfig,
+          async () => {
+            return (token as unknown as typeof obj).key;
+          }
+        ),
+        outputs: { table: table.tableArn },
+      };
+    },
+    (outputs) => {
+      return outputs.table;
+    }
+  );
+
+  test(
+    "serialize token with lazy should fail",
```

Comment:
Why should it fail? Cant we resolve the value?

Reply:
We can, though I am concerned that we'd be resolving the lazy too early.

Lazy relies on being resolved at the last moment to catch anything that may change with its contents (ex: adding more GSIs). However we current start the lambda serialization immediately and may end up with a race condition.

Need to test more.

Reply:
Maybe we should start the closure serializer in `serialize` instead of constructor? That way the construct tree is fully instantiated at least.

Reply:
Was thinking about that...

I don't think serialize would work.

* At bundling happens at the beginning
* The serialize hook seems to be closed off, get an error about using validate
* There is no way to await a promise. 

One thought was to start the promises at the end of construct creation, but before synth, but that would require an explicit command be added to start compilation and would take longer to complete. 

`await Functionless.compileFunctions`

Reply:
What do you mean the serialize hook seems closed off?

Reply:
Trying to use onSynthesize threw an error and said to use validate. 

Either way... The async issue is not something we can bypass easily.